containers/datalab/run-with-gce.sh

```diff
@@ -83,8 +80,33 @@ fi
 
 echo "Looking up gateway VM for ${USER_EMAIL}..."
 
-USER_HASH=`echo "${USER_EMAIL}" | openssl sha1 -r | cut -d ' ' -f 1`
-INSTANCE_PREFIX="datalab-gateway-${USER_HASH:0:12}"
+# We would like to give the VM a friendly name based on the user, but VM names
+# can only contain letters, numbers, and hyphens, whereas email addresses can
+# contain any characters.
+#
+# We resolve this conflict by trying a simple escaping of the email address,
+# and if that does not produce a valid instance name, then we just hash the
+# email address and use the hash.
+#
+# The escaping we do is based on a 1:1 mapping from '[-@.a-zA-Z0-9]+' to
+# '[-a-zA-Z0-9]+'. Since this mapping is 1:1, we ensure that the uniqueness
+# of email addresses is maintained. There is additionally a length limit
+# of 63 characters for instance names, so we do not use the escaped email
+# address if it is longer than 32 characters.
+EMAIL_WITH_HYPHENS_ESCAPED=`echo "${USER_EMAIL}" | sed -e 's/-/-h-/g'`
+EMAIL_WITH_AT_ESCAPED=`echo "${EMAIL_WITH_HYPHENS_ESCAPED}" | sed -e 's/@/-at-/g'`
+EMAIL_WITH_DOT_ESCAPED=`echo "${EMAIL_WITH_AT_ESCAPED}" | sed -e 's/\./-dot-/g'`
+ESCAPED_USER_EMAIL=`echo "${EMAIL_WITH_DOT_ESCAPED}" | sed -e 's/[^-a-zA-Z0-9]//g'`
+TRUNCATED_USER_EMAIL="${ESCAPED_USER_EMAIL:0:32}"
+if [[ -z "${ESCAPED_USER_EMAIL}" || "${ESCAPED_USER_EMAIL}" != "${TRUNCATED_USER_EMAIL}" ]]; then
```

Comment:
> Since we only use TRUNCATED_USER_EMAIL when it is equal to ESCAPED_USER_EMAIL, maybe we can just have: 
> if [[ -z "${ESCAPED_USER_EMAIL}" || {#ESCAPED_USER_EMAIL} > 32 ]]; then
and no need to truncate the email?

Nice! I didn't know you could do that in bash, so thanks for the suggestion.